### PR TITLE
Add support for optional size threshold to archive log file on startup

### PR DESCRIPTION
**Pull request by Sam13**
With that feature it's possible to archive the old file on startup when a certain size threshold is reached.

By enabling this you can ensure that the current application session is always in the same file but still have some kind of size based archival. Could also be comined with other archival options.

### CODE REVIEW

**Comment by 304NotModified** [comment]
Thanks!

**Comment by snakefoot** [comment]
Could this code:

```c#
var length = _fileAppenderCache.GetFileLength(fileName);
if (ArchiveOldFileOnStartup && length.HasValue && length.Value > ArchiveOldFileOnStartupAboveSize)
{
    // Do some stuff
}
```

Be changed into this code (Avoids using _fileAppenderCache unless necessary):

```c#
if (ArchiveOldFileOnStartup)
{
   if (ArchiveOldFileOnStartupAboveSize <= 0 || ArchiveOldFileOnStartupAboveSize < _fileAppenderCache.GetFileLength(fileName))
   {
        // Do some stuff
   }
}
```

Could you create an isolated unit-test instead of piggy-backing on an existing one? Let the unit-test-name include the setting `ArchiveOldFileOnStartupAboveSize`.

**Comment by Sam13** [comment]
> Be changed into this code (Avoids using _fileAppenderCache unless necessary):

Ok will do

> Could you create an isolated unit-test instead of piggy-backing on an existing one? Let the new unit-test-name include the setting `ArchiveOldFileOnStartupAboveSize`.

That's what I initially wanted to do but I thought I'd conform to the existing style

**Comment by snakefoot** [comment]
Please extract the new condition logic into a method called ShouldArchiveOldFileOnStartup


There is a lot of technical debt in the NLog unit tests for the FileTarget. But it is hard work to cleanup. But will be easier if not piling more on top :)


Sent from my Sony Xperia


---- Sam13 wrote ----


Be changed into this code (Avoids using _fileAppenderCache unless necessary):

Ok will do

Could you create an isolated unit-test instead of piggy-backing on an existing one? Let the new unit-test-name include the setting ArchiveOldFileOnStartupAboveSize.

That's what I initially wanted to do but I thought I'd conform to the existing style

**Comment by Sam13** [comment]
> There is a lot of technical debt in the NLog unit tests for the FileTarget. But it is hard work to cleanup. But will be easier if not piling more on top :)

I fully agree - all findings addressed

**Comment by Sam13** [comment]
https://travis-ci.org/NLog/NLog/builds/634731645?utm_source=github_status&utm_medium=notification

```
NLog.UnitTests.Targets.PlainFileTargetTests.DateArchive_SkipPeriod(timeKind: Local, archivePeriod: Hour, includeDateInLogFilePath: True, includeSequenceInArchive: True) [FAIL]
      Assert.Equal() Failure
                          ↓ (pos 10)
      Expected: 202001091310.0.txt
      Actual:   202001091309.0.txt
                          ↑ (pos 10)
```

seems to be a flaky unit test and not related to my changes

**Comment by 304NotModified** [comment]
FYI Restarted Travis

**Comment by 304NotModified** [comment]
Looks good to me :)   agreed @snakefoot ?

**Comment by snakefoot** [comment]
I would probably have made it so setting ArchiveOldFileOnStartupAboveSize was enough alone. But since I will not be using it, then it is not that important. It looks good to me. Maybe include in NLog 4.7?


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


Looks good to me :) agreed @snakefoot<https://github.com/snakefoot> ?

**Comment by Sam13** [comment]
> 
> I would probably have made it so setting ArchiveOldFileOnStartupAboveSize was enough alone. 

I thought it's more intuitive to combine with the flag _ArchiveOldFileOnStartup_. Otherwise you can configure the behavior to archive on startup with two different options (setting flag to true or setting threshold >= 0) which may cause confusion.

**Comment by snakefoot** [comment]
> I thought it's more intuitive to combine with the flag ArchiveOldFileOnStartup. 

We also have "archiveAboveSize" and "archiveEvery", and they work right away without first having configured an additional parameter called "ArchiveEnabled". But both our opinions are valid :)

**Comment by 304NotModified** [comment]
> I thought it's more intuitive to combine with the flag _ArchiveOldFileOnStartup_. Otherwise you can configure the behavior to archive on startup with two different options (setting flag to true or setting threshold >= 0) which may cause confusion.


I'm afraid it also confusing that ArchiveOldFileOnStartupAboveSize =2 is confusing, at it won't work without ArchiveOldFileOnStartup.

proposal, if we won't set ArchiveOldFileOnStartup, and we do set ArchiveOldFileOnStartupAboveSize , then the default for ArchiveOldFileOnStartup will be true (instead of false).

Or is this proposal making things to complicated?

**Comment by Sam13** [comment]
> I'm afraid it also confusing that ArchiveOldFileOnStartupAboveSize =2 is confusing, at it won't work without ArchiveOldFileOnStartup.
 
You are right but I think it does not matter with proper documentation.

Another option I can think of is to leave _ArchiveOldFileOnStartup_ away since this is basically the same than _ArchiveOldFileOnStartupAboveSize = 0_
Disabling _ArchiveOldFileOnStartup_  then means to unset this option.

Since you are the maintainers please define on how it should behave and I'll try to implement that.

File: src/NLog/Targets/FileTarget.cs [hunk 1]
```diff
@@ -2183,6 +2196,28 @@ private void WriteFooter(string fileName)
...
 
+        /// <summary>
+        /// Decision logic whether to archive logfile on startup.
+        /// <see cref="ArchiveOldFileOnStartup"/> and <see cref="ArchiveOldFileOnStartupAboveSize"/> properties.
+        /// </summary>
+        /// <param name="fileName">File name to be written.</param>
+        /// <returns>Decision whether to archive or not.</returns>
+        private bool ShouldArchiveOldFileOnStartup(string fileName)
+        {
+            if (!ArchiveOldFileOnStartup)
```
Comment (snakefoot) [inline src/NLog/Targets/FileTarget.cs]: Change this to `if (!ArchiveOldFileOnStartup && ArchiveOldFileOnStartupAboveSize <= 0)`

Comment (304NotModified) [reply src/NLog/Targets/FileTarget.cs]: isn't that conflicting with https://github.com/NLog/NLog/pull/3755/files/97b0d69e88637e378fba313af3ac42a3a0664e10#diff-ed0a3fc25ad7408a5b63501fa77dbe69R2212-R2215 ?

Comment (304NotModified) [reply src/NLog/Targets/FileTarget.cs]: (PS @snakefoot I could edit the PR also from the Github GUI, I think that also counts for you)

Comment (snakefoot) [reply src/NLog/Targets/FileTarget.cs]: I guess the pretty version would be this:

```c#
        private bool ShouldArchiveOldFileOnStartup(string fileName)
        {
            // No size threshold specified
            if (ArchiveOldFileOnStartupAboveSize <= 0)
            {
                 return ArchiveOldFileOnStartup;
            } 
```

Comment (304NotModified) [reply src/NLog/Targets/FileTarget.cs]: it's the same behavior, just another version how to write, isn't?

I find it debatable which one is more readable. In terms of performance is no  (significant) difference.

Comment (Sam13) [reply src/NLog/Targets/FileTarget.cs]: > I guess the pretty version would be this:
> 
> ```cs
>         private bool ShouldArchiveOldFileOnStartup(string fileName)
>         {
>             // No size threshold specified
>             if (ArchiveOldFileOnStartupAboveSize <= 0)
>             {
>                  return ArchiveOldFileOnStartup;
>             } 
> ```

Pretty but more complicated in my opinion. My version is straight forward and easy to understand that's why I did it with multiple return statements.

But as I said. Tell me what you want and I'll change it.

Comment (304NotModified) [reply src/NLog/Targets/FileTarget.cs]: No need to change IMO

**Comment by snakefoot** [comment]
Remember to rebase on master to get it on NLog 4.7

**Comment by Sam13** [comment]
> 
> 
> Remember to rebase on master to get it on NLog 4.7
I assumed that all new features land on _dev_ first?
Should I rebase on master and change the PR's base to master?

**Comment by snakefoot** [comment]
DEV branch is targeting NLog ver. 5. So small features and bugfixes are committed into master and merged back to dev on release.


Sent from my Sony Xperia


---- Sam13 wrote ----


Remember to rebase on master to get it on NLog 4.7
I assumed that all new features land on dev first?
Should I rebase on master and change the PR's base to master?

**Comment by Sam13** [comment]
Change base branch to master

**Comment by 304NotModified** [comment]
close/reopen to trigger CI

**Comment by 304NotModified** [comment]
> You are right but I think every combination with two properties is not that complicated with proper documentation.
> 
> Another option I can think of is to leave _ArchiveOldFileOnStartup_ away since this is basically the same than _ArchiveOldFileOnStartupAboveSize = 0_
> Disabling _ArchiveOldFileOnStartup_ then means to unset this option.
> 
> Since you are the maintainers please define on how it should behave and I'll try to implement that.

Oops this is still open. It still propose: https://github.com/NLog/NLog/pull/3755#issuecomment-576874258

**Comment by snakefoot** [comment]
> Oops this is still open. It still propose: #3755 (comment)

You just said that the PR was [fine as it is](https://github.com/NLog/NLog/pull/3755#discussion_r370291993):

**Comment by 304NotModified** [comment]
Well that was a refactor proposal and not a functional change?

**Comment by snakefoot** [comment]
> Well that was a refactor proposal and not a functional change?

No it was a functional change, that implemented the proposed change

**Comment by 304NotModified** [comment]
Was it the code for https://github.com/NLog/NLog/pull/3755#issuecomment-576874258? I don't see that.

**Comment by snakefoot** [comment]
Guess my proposal was too complex then :)

**Comment by 304NotModified** [comment]
I changed it to https://github.com/NLog/NLog/pull/3755#issuecomment-576874258

Also I made the FIleApenderCache mockable (we should really do more mocking). But a bit afraid of the performance concequence. (interface call vs class class). What do you think @snakefoot ?

**Comment by snakefoot** [comment]
Think the introduction of `_archiveOldFileOnStartup` makes it even more confusing. Think my [suggestion](https://github.com/NLog/NLog/pull/3755#discussion_r369845004) is better.

**Comment by 304NotModified** [comment]
No opinion on the IFileApenderCache?

**Comment by snakefoot** [comment]
Think it is the wrong place to mock. Lots of action happening in the file appender cache. Maybe a IFileAppender interface and then inject a factory that returns a mock of the interface


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


No opinion on the IFileApenderCache?

**Comment by 304NotModified** [comment]
> Think it is the wrong place to mock. Lots of action happening in the file appender cache.

I also checked that, but the appendar cache is doing File stuff (FileInfo) and that is exactly what needs to be "mocked out".

**Comment by 304NotModified** [comment]
@snakefoot I you don't think we should performance test this, then I like to merge it. So please let me know if you're OK with that :)

**Comment by snakefoot** [comment]
I don't use any of the features. But I still think your latest change makes the code even more confusing. Anyway it will not hurt performance.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


@snakefoot<https://github.com/snakefoot> I you don't think we should performance test this, then I like to merge it. So please let me know if you're OK with that :)

**Comment by snakefoot** [comment]
Now if having configured archive at startup and archive at startup when above size. Then it will always archive independent of size. This will be confusing. Again my propasal is better.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


@snakefoot<https://github.com/snakefoot> I you don't think we should performance test this, then I like to merge it. So please let me know if you're OK with that :)

**Comment by snakefoot** [comment]
I think there should be a IFileSystem interface that FileTarget should share with FileAppenderCache.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


Think it is the wrong place to mock. Lots of action happening in the file appender cache.

I also checked that, but the appendar cache is doing File stuff (FileInfo) and that is exactly what needs to be "mocked out".

**Comment by snakefoot** [comment]
But maybe this PR is not the right PR for rewriting FileTarget to support mocking.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


Think it is the wrong place to mock. Lots of action happening in the file appender cache.

I also checked that, but the appendar cache is doing File stuff (FileInfo) and that is exactly what needs to be "mocked out".

**Comment by 304NotModified** [comment]
> But maybe this PR is not the right PR for rewriting FileTarget to support mocking.

Maybe, but we need tests. 

> Now if having configured archive at startup and archive at startup when above size. Then it will always archive independent of size. This will be confusing.

True, will fix that

**Comment by 304NotModified** [comment]
> Now if having configured archive at startup and archive at startup when above size. Then it will always archive independent of size. This will be confusing.

Fixed!

> But maybe this PR is not the right PR for rewriting FileTarget to support mocking.

We could refactor this in another PR. It's all internal.

**Comment by snakefoot** [comment]
Find it strange to write mocking code that you know upfront needs to be rewritten right after.  The mock should cover all file-access and just half.


But yes now archive on startup when above size will work as expected.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


But maybe this PR is not the right PR for rewriting FileTarget to support mocking.

Maybe, but we need tests.

Now if having configured archive at startup and archive at startup when above size. Then it will always archive independent of size. This will be confusing.

True, will fix that

**Comment by Sam13** [comment]
Any ETA when this gets merged and released?

**Comment by 304NotModified** [comment]
Will try to merge this weekend and release next week. Release is a bit dependent of the docs (if it's up-to-date already)

**Comment by 304NotModified** [comment]
> Find it strange to write mocking code that you know upfront needs to be rewritten right after.

Well I have no time now to make it perfect. 

So I will merge this for now.

**Pull request merged**